pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,489 @@
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	log "github.com/sirupsen/logrus"
+	rv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/storage/names"
+)
+
+var ObjectTypeNotFoundInBundleError = errors.New("object kind not found in bundle")
+
+type synthesize struct {
+	bundleDir             string
+	metadataDirName       string
+	nonInferableCSV       *NonInferableCSV
+	unstructuredObjectMap map[string][]unstructured.Unstructured // map[kind]: *array of unstructured objects
+	existingCSVPath       string
+	synthesizedCSV        *operatorsv1alpha1.ClusterServiceVersion
+	apiDescriptionMap     map[GroupVersionKind]*ApiDescription
+	logger                *log.Entry
+	dependencies          *registry.DependenciesFile
+	scheme                *runtime.Scheme
+}
+
+// SynthesizeFromDirectory initialize synthesize, registry and load all manifests needed for the CSV from the bundle.
+func SynthesizeFromDirectory(bundleDir, metadataDirName string) (*synthesize, error) {
+	_, err := ioutil.ReadDir(bundleDir)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &synthesize{
+		bundleDir:             bundleDir,
+		metadataDirName:       metadataDirName,
+		nonInferableCSV:       &NonInferableCSV{},
+		unstructuredObjectMap: map[string][]unstructured.Unstructured{},
+		logger:                log.WithFields(log.Fields{"type": "synthesize"}),
+		apiDescriptionMap:     map[GroupVersionKind]*ApiDescription{},
+		scheme:                runtime.NewScheme(),
+	}
+
+	registerGVK(s.scheme)
+
+	if err := s.loadUnstructuredObjects(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *synthesize) GenerateCSV() (*operatorsv1alpha1.ClusterServiceVersion, error) {
+	if err := s.generateCsv(); err != nil {
+		if err == CsvInferenceFileNotFoundError && s.existingCSVPath == "" {
+			return nil, fmt.Errorf("a bundle should contain either a CSV or non-inferable CSV information in the %s directory", s.metadataDirName)
+		}
+		return nil, err
+	}
+
+	return s.synthesizedCSV, nil
+}
+
+func (s *synthesize) generateCsv() error {
+	csvBasics, err := s.getNonInferableCSV()
+	if err != nil {
+		return err
+	}
+
+	installStrategy, err := s.getInstallStrategy()
+	if err != nil {
+		return err
+	}
+
+	crdDefinitions, apiServiceDefinitions, err := s.getAPIs()
+	if err != nil {
+		return err
+	}
+
+	s.synthesizedCSV = &operatorsv1alpha1.ClusterServiceVersion{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: operatorsv1alpha1.ClusterServiceVersionAPIVersion,
+			Kind:       CSVKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: csvBasics.Name,
+		},
+		Spec: operatorsv1alpha1.ClusterServiceVersionSpec{
```

Comment:
Are webhook definitions going in a followup?

Reply:
Yes, I know it is going to be a follow up, but i dont know when exactly.